src/Interpreters/ConcurrentHashJoin.h

```diff
@@ -80,6 +80,11 @@ class ConcurrentHashJoin : public IJoin
         return std::make_shared<ConcurrentHashJoin>(context, table_join_, slots, right_sample_block_, stats_collecting_params);
     }
 
+    std::shared_ptr<IJoin> cloneNoParallel(const std::shared_ptr<TableJoin> & table_join_, const Block &, const Block & right_sample_block_) const override
```

Comment:
This is needed to automatically replace parallel_hash to hash with sharding, cause hash is more performant in this case.

Probably, the sharding optimization can be done on top of the JoinLogical step to choose the algorithm after the sharding. However, currently it works fine with the physical join step.

Reply:
The downside that the change is not visible in EXPLAIN